src/Dialect/ONNX/DialectBuilder.cpp

```diff
@@ -43,6 +43,16 @@ Value OnnxBuilder::add(Value A, Value B) const {
   return createOpAndInferShapes<ONNXAddOp>(toTensor(A), toTensor(B));
 }
 
+Value OnnxBuilder::cast(
+    Type outputType, Value input, IntegerAttr saturate, TypeAttr to) const {
+  if (outputType.cast<ShapedType>().hasRank()) {
+    return cast(input, saturate, to);
+  } else {
+    Type resultType = UnrankedTensorType::get(to.getValue());
+    return b().create<ONNXCastOp>(loc(), resultType, input, saturate, to);
+  }
```

Comment:
It's a bit confusing here to know the meaming of `outputType`. `outputType` often means the type of the output, so we can use it directly, say `return b().create<ONNXCastOp>(loc(), outputType, input, saturate, to);`. The logic of constructing `resultType` should be outside of this function.

I suggest to define this function as follows:
```
Value OnnxBuilder::cast(
    Type outputType, Value input, IntegerAttr saturate, TypeAttr to) const {
    return createTypedOpAndInferShapes<ONNXCastOp>(outputType, input, saturate, to);
}
```

Reply:
@tungld Thanks for your comments!
I tried this before, but when inferring shape here using `createTypedOpAndInferShapes`, rewriting CastLike op to Cast op failed. I think this is because tensor type mismatches since output type of CastLike op is unranked tensor type, but output of Cast op is ranked tensor type with unknown dimension after inferring shape. Is this strange behavior?

Reply:
I see. So it means we have two cases: use outType as it is and inferShape. 

How about adding a flag `inferShape` (set to true by default) to the function? 

If `inferShape == true` we will return  `createTypedOpAndInferShapes<ONNXCastOp>(outputType, input, saturate, to);` otherwise `b().create<ONNXCastOp>(loc(), outputType, input, saturate, to);`

Reply:
Done. Thanks!